keychain.h

```diff
@@ -0,0 +1,125 @@
+#include <assert.h>
+
+#include <optional>
+#include <string>
+#include <variant>
+
+namespace keychain {
+
+class Keychain {
+   public:
+    struct Error {
+        std::string message;
+    };
+    using Success = std::monostate;
+    using Password = std::string;
+    using Result = std::variant<Success, Password, Error>;
+
+    static Result getPassword(const std::string &service,
+                              const std::string &user);
+    static Result setPassword(const std::string &service,
+                              const std::string &user,
+                              const std::string &password);
+    static Result deletePassword(const std::string &service,
+                                 const std::string &user);
+
+   public:
+    /*! \brief Construct a Keychain object
+     *
+     * \param package Package identifier corresponding to the format
+     *                com.example.application. On MacOS this identifier is pre-
+     *                pended to the service name and is visible to users. On
+     *                Linux it is used as the schema name and is normally not
+     *                visible to users. On Windows this parameter is ignored.
+     */
+    Keychain(const std::string &package);
+
+    /*! \brief Get a password
+     *
+     * \param service Identifier for the service whose passwords should be
+     * stored or updated.
+     *
+     * \param username Username or account name that the stored or updated
+     *                 passwords belong to.
+     */
+    std::optional<std::string> getPassword(const std::string &service,
+                                           const std::string &username);
+
+    /*! \brief Update a password
+     *
+     * \param password The password to be stored.
+     *
+     * \param service Identifier for the service whose passwords should be
+     * stored or updated.
+     *
+     * \param username Username or account name that the stored or updated
+     *                 passwords belong to.
+     */
+    bool setPassword(const std::string &password, const std::string &service,
+                     const std::string &username);
+
+    /*! \brief Delete a password
+     *
+     * \param service Identifier for the service whose passwords should be
+     * stored or updated.
+     *
+     * \param username Username or account name that the stored or updated
+     *                 passwords belong to.
+     */
+    bool deletePassword(const std::string &service,
+                        const std::string &username);
+
+    /*! \brief Get the last error, if any
+     *
+     * Return the error resulting from the most recent call to getPassword,
+     * addPassword, or deletePassword. If the last call to any of the functions
+     * did not result in an error, std::nullopt is returned.
+     */
+    std::optional<std::string> lastError() const;
+
+   protected:
+    void clearError();
+
+    template <typename F>
+    bool updatePassword(F &&updateFun);
```

Comment:
Both `setPassword` and `deletePassword` need to do the same result handling to convert the static function's `Result` to the `bool` and setting `lastError`. In addition, there's the `clearError` and `assert` part.
Depending on what we decide that the c'tors and get/set/delete functions' parameters should be, this could be obsolete in any case.